bpo-37424: Avoid a hang in subprocess.run timeout output capture

gpshead opened the pull request:
Fixes a possible hang when using a timeout on `subprocess.run()` while
capturing output.  If the child process spawned its own children or otherwise
connected its stdout or stderr handles with another process, we could hang
after the timeout was reached and our child was killed when attempting to read
final output from the pipes.

<!-- issue-number: [bpo-37424](https://bugs.python.org/issue37424) -->
https://bugs.python.org/issue37424
<!-- /issue-number -->

gpshead committed:
bpo-37424: Have subprocess use killpg for some timeout kills.

When shell=True the user is often launching a process tree, if we
kill only the shell and it hasn't set itself up as a pgrp leader,
the grandchild processes still run and may have our output handles
open.  Leading us to wait forever when we should be timing out.

This Adds a fractional timeout on the cleanup communicate phase
_and_ will use killpg on the child process group _if_ it is
different from our own as it should be when start_new_session=True
was used.

gpshead committed:
Adds a test for the no session scenario (common).

giampaolo commented on Lib/subprocess.py:
```diff
@@ -505,6 +511,11 @@ def run(*popenargs,
 
+def _get_cleanup_timeout(timeout):  # For test injection.
+    if timeout is not None:
+        return timeout/20
```
why timeout / 20? (instead of just returning original timeout)

gpshead replied:
The original full timeout has already elapsed and expired, SIGKILL was
sent, we don't want to wait much longer.

vstinner commented on Lib/subprocess.py:
```diff
@@ -491,7 +491,13 @@ def run(*popenargs,
             stdout, stderr = process.communicate(input, timeout=timeout)
         except TimeoutExpired:
             process.kill()
```
Why don't we close all pipes here? Closing pipes + .wait() should be enough, no? SIGKILL cannot be ignored on Unix (I don't know for Windows).

gpshead replied:
Agreed this implementation is too complex.  That sounds tempting and simple.  Trying that, I found that it caused the capture_output_timeout tests to fail as we no longer collected the output and attached it to the Timeout exception.  But that led to some code inspection... we're already gathering the output within the original _communicate() on POSIX and can attach everything we have in the _original_ Timeout exception so that we don't need to call communicate() again after kill() at all in this path.

Windows collects output in a different manner so it still needs it, but this bug is POSIX only so I've reworked this to simplify it.  The communicate call for output gathering becomes Windows only.  I'm relying on the CI systems to do my Windows testing.

gpshead committed:
Simplify this a lot, do not use magic process groups.

Do not have a post-timeout sub-timeout for output collection, we've already
got all of that output within POSIX's _communicate() method as it is.

gpshead commented on Lib/subprocess.py:
```diff
@@ -1910,7 +1933,15 @@ def send_signal(self, sig):
             # Skip signalling a process that we know has already died.
             if self.returncode is None:
-                os.kill(self.pid, sig)
+                if self.__signal_process_group:
```
Should I also undo the addition of this magic when shell=True process group killing logic?

The bug was about a hang after the timeout should've expired.  That hang will no longer happen and we will return.  But this does potentially leave the shell's child processes running on some posix platforms and child process configurations.  Likely not what the caller ever intended.

People running on stable Python releases who need to ensure the process group's children have died already need to do their own post-timeout getpgid() + killpg() logic dance similar to what is in here no matter what.  _This_ logic feels like a feature that I'd do for 3.8+ only.

Adding explicit process group killing features to the kill APIs as some of the dup'ed issues have desired is also squarely a feature request, likely a good idea, but belongs in its own set of PRs as it won't be backported as a bugfix.

gpshead replied:
okay, i've convinced myself.  process group related logic belongs in its own PR.  people need to manually specify start_new_session for this logic to be meaningful regardless.  and for uniformity there's a question of what makes sense on windows as it has its own notion of process groups but *not* a start_new_session concept.

gpshead committed:
Remove the magic shell pgrp detection killpg logic.

that belongs in its own PR if at all.

gpshead committed:
Remove session pgrp logic related test.

it wasn't actually testing what it described as it never checked
the grandchild process explicitly.  redo this better in a later PR.

gpshead committed:
blurb

Yhg1s commented on Lib/subprocess.py:
```diff
@@ -489,11 +489,20 @@ def run(*popenargs,
+                exc.stdout, exc.stderr = process.communicate()
+            else:
+                # POSIX _communicate already populated the output so
+                # far into the TimeourExpired exception.
```
Timeour -> Timeout

Yhg1s commented on Lib/test/test_subprocess.py:
```diff
@@ -1564,6 +1565,22 @@ def test_stderr_with_capture_output_arg(self):
+            stacks = traceback.format_exc()  # assertRaises doesn't give this.
+        else:
+            self.fail("TimeoutExpired not raised.")
+        self.assertLess(after_secs - before_secs, 1.5,
```
This feels like it would be sensitive to heavily loaded machines... Maybe add a comment to it to make it easier to identify if that is the case?

gpshead committed:
Typo in comment, add a test comment.
